qri/dataset.py

```diff
@@ -60,48 +59,76 @@ def __repr__(self):
         return 'Commit(%s)' % r
 
 
+def is_from_list(obj):
+    version_info_fields = ['bodySize', 'bodyRows', 'bodyFormat', 'numErrors', 'commitTime']
+    return bool(set(obj) & set(version_info_fields))
+
+
 class Dataset(object):
     def __init__(self, obj):
         # Fields that are always present
         set_fields(self, obj, ['username', 'name', 'profileID', 'path'])
         if self.username is None and 'peername' in obj:
             self.username = obj.get('peername')
-        # Fields returned by `get` commands
-        set_fields(self, obj, ['bodyPath', 'previousPath'])
-        # Version info
-        self.versionInfo = self._build_version_info(obj)
+
+        self._is_populated = False
+        if not is_from_list(obj):
+            self._populate(obj)
+
+    def _ensure_populated(self):
+        if self._is_populated:
+            return
+        ref = dsref.Ref(self.username, self.name)
+        self._populate(loader.instance().get_dataset_object(ref))
+
+
+    # TODO - write tests
+    def _populate(self, obj):
+        self.body_path_value = obj.get('bodyPath')
+        self.previous_path_value = obj.get('previousPath')
+
         # Subcomponents
         self.commit_component = Commit(obj.get('commit'))
         self.meta_component = Meta(obj.get('meta'))
         self.readme_component = Readme(obj.get('readme'))
         self.structure_component = Structure(obj.get('structure'))
```

Comment:
So here's where it could have gotten complicated. Supposing we want the user to be able to say `ds.structure.entries` without populating `ds` (since it's in `versionInfo`), but once they say `ds.structure.checksum`, it populates. I'm going to question here whether it's worth it (assuming I didn't misunderstand your suggestion in the first place).

To implement, it would require the populating logic to be part of `Structure`, since (to my knowledge) `Dataset` can't have something like a `@property` that operates on `structure.checksum`. or `structure.*`. That would mean that in the case of `versionInfo` present, I'd have to do something like:

```
self.structure_component = Structure(obj, self)
```

And the constructor of `Structure` would have:

```
def __init__(self, obj, ds=None):
    if is_from_list(obj):
        self.entries = obj.get('bodyRows')
        # and so on for every versionInfo field

        self.parent = ds
    else:
        set_fields(...obj...) # Initialize from `get` as usual

@property
def checksum(self):
    self.parent.ensure_populated()
    return self.checksum

# and so on for every field of Structure
```

My question is, is there much gain to be gained for the trouble, future maintenance costs as there are changes to fields, and the synchronization issues I brought up earlier?